src/riscv/lib/src/jit/builder/control_flow_graph.rs

```diff
@@ -340,6 +364,95 @@ where
+        let mut seq_max_budget = 0;
+
+        let mut work_stack = self
+            .graph
+            .outgoing_outcomes(SourceInstrLoc::Entry)
+            .iter()
+            .filter_map(|outcome| {
+                self.outcomes[*outcome]
+                    .to()
+                    .as_internal()
+                    .map(|&node_id| (node_id, self.sorted_outgoings(node_id).collect_vec()))
+            })
+            .collect_vec();
+
+        while let Some((node_id, child_outcomes)) = work_stack.pop() {
+            let mut max_budget = 1;
+            for child_outcome in &child_outcomes {
+                if bc_locations[**child_outcome].data().is_some() {
+                    // Budget check edge, budget is 1.
+                    max_budget = max_budget.max(1);
+                    continue;
+                }
+
+                let child_node_id = match self.outcomes[**child_outcome].to() {
+                    TargetInstrLoc::Internal(id) => id,
+                    TargetInstrLoc::Exit => {
+                        max_budget = max_budget.max(1);
+                        continue;
+                    }
+                };
+
+                match budgets[child_node_id] {
+                    Some(child_budget) => {
+                        max_budget = max_budget.max(1 + child_budget);
+                    }
+                    None => {
+                        // Child budget not set, need to process child first.
+                        work_stack.push((node_id, child_outcomes));
+                        work_stack.push((
+                            child_node_id,
+                            self.sorted_outgoings(child_node_id).collect_vec(),
+                        ));
+                        break;
+                    }
+                }
+            }
+
+            budgets[node_id] = Some(max_budget);
+            seq_max_budget = seq_max_budget.max(max_budget);
+        }
+
+        (seq_max_budget, budgets)
+    }
+
+    /// Fill in budget check values at the given budget check locations.
+    ///
+    /// Return the maximum budget required for the entire sequence along with
+    /// the filled budget checks.
+    pub fn fill_budget_check_values(
+        &self,
+        bc_locations: &OutcomeMap<Option<BudgetCheckLoc<'info, T>>>,
+    ) -> (usize, OutcomeMap<Option<BudgetCheck<'_, T>>>) {
+        let (seq_max_budget, budgets) = self.annotate_budgets(bc_locations);
+
+        // Now fill in the budget check values.
+        let budget_checks = bc_locations.map(|outcome_id, outcome| {
+            outcome.data().as_ref().and_then(|bc_loc| {
+                self.outcomes[outcome_id]
+                    .to()
+                    .as_internal()
+                    .and_then(|node_id| budgets[*node_id])
+                    .map(|target| BudgetCheck {
+                        budget: target,
+                        edge: bc_loc.edge(),
+                    })
+            })
```

Comment:
Same with this block. Let's flatten the structure to make it approachable.

```suggestion
            let bc_loc = outcome.data().as_ref()?;
            let &node_id = self.outcomes[outcome_id].to().as_internal()?;
            let budget_check = BudgetCheck {
                budget: budgets[node_id]?,
                edge: bc_loc.edge(),
            };
            Some(budget_check)

```

Reply:
thank you. done [here](https://github.com/tezos/riscv-pvm/compare/ec0e39b26dea1e42ee2fac7372d4f0288b41cae1..fde1fd0a02cec17671bdeef11b4f4c38cdedd7cc)